components/remote_settings/src/client.rs

```diff
@@ -0,0 +1,551 @@
+/// Methods defined on this will fetch data from
+/// <base_url>/v1/buckets/<bucket_name>/collections/<collection_name>/
+pub struct Client {
+    pub(crate) base_url: Url,
+    pub(crate) bucket_name: String,
+    pub(crate) collection_name: String,
+    pub(crate) remote_state: Mutex<RemoteState>,
+}
+
+impl Client {
+    /// Create a new [Client] with properties matching config.
+    /// During construction, the [Client] will fetch the base attachments
+    /// URL from the server's metadata, assuming attachments are enabled for
+    /// the server.
+    pub fn new(config: RemoteSettingsConfig) -> Result<Self> {
+        let server_url = config
+            .server_url
+            .unwrap_or_else(|| String::from("https://firefox.settings.services.mozilla.com"));
+        let bucket_name = config.bucket_name.unwrap_or_else(|| String::from("main"));
+        let base_url = Url::parse(&server_url)?;
+
+        Ok(Self {
+            base_url,
+            bucket_name,
+            collection_name: config.collection_name,
+            remote_state: Default::default(),
+        })
+    }
+
+    /// Fetches all records for a collection that can be found in the server,
+    /// bucket, and collection defined by the [ClientConfig] used to generate
+    /// this [Client].
+    pub fn get_records(&self) -> Result<RemoteSettingsResponse> {
+        let path = format!(
+            "v1/buckets/{}/collections/{}/records",
+            &self.bucket_name, &self.collection_name
+        );
+        let resp = self.make_request(path)?;
+        let records = resp.json::<RecordsResponse>()?.data;
+        let last_modified = resp.headers.get_as("etag").unwrap()?;
+        Ok(RemoteSettingsResponse {
+            records,
+            last_modified,
+        })
+    }
+
+    /// Fetches all records for a collection that can be found in the server,
+    /// bucket, and collection defined by the [ClientConfig] used to generate
+    /// this [Client]. This function will return the raw network [Response].
+    pub fn get_records_raw(&self) -> Result<Response> {
+        let path = format!(
+            "v1/buckets/{}/collections/{}/records",
+            &self.bucket_name, &self.collection_name
+        );
+        self.make_request(path)
+    }
+
+    /// Fetches all records that have been published since provided timestamp
+    /// for a collection that can be found in the server, bucket, and
+    /// collection defined by the [ClientConfig] used to generate this [Client].
+    pub fn get_records_since(&self, timestamp: u64) -> Result<RemoteSettingsResponse> {
+        let path = format!(
+            "v1/buckets/{}/collections/{}/records?_since={}",
+            &self.bucket_name, &self.collection_name, timestamp
+        );
+        let resp = self.make_request(path)?;
+        let records = resp.json::<RecordsResponse>()?.data;
+        let last_modified = resp.headers.get_as("etag").unwrap()?;
```

Comment:
There is a `ResponseError(String)` that seems suitable, so the cleaner story is probably just `.ok_or_else(|| RemoteSettingsError::ResponseError("no etag in response".into())?;` or something?  (OTOH though, IIRC this unwrap already existed and has forever - no etag does mean something very unexpected, and panics in our mobile clients aren't as catastrophic as they are on desktop)

Reply:
This is a little further complicated because `get_as` returns an `Option<Result<....>>` and not just `Result<...>`. I matched on that pretty explicitly, but please let me know if there's a way to simplify it down some

Reply:
Nice, thanks! You can tighten it up a bit with:

```rs
let last_modified = resp.headers.get_as("etag").ok_or_else(|| RemoteSettingsError::ResponseError("no etag header".into()))??;
```

`ok_or_else` turns your `Option<Result<...>>` into a `Result<Result<...>>`, then the first `?` unwraps the outer one, and the second `?` unwraps the inner one.

Reply:
TY, that's much better 🙏